Grammar issues in memory-optimizations document (#2370)
RedisLabs/redislabs-docs: content/ri/memory-optimizations.md

alwaysopen: false
categories: ["RI"]
---
RedisInsight can monitor the health of your databases. Here are some tips to increase memory efficiency in Redis.

To get the best performance from your databases, ensure you are using the latest version of [Redis](https://redis.io).

## Developer best practices

### Avoid dynamic Lua script

Refrain from generating dynamic scripts, which can cause your Lua cache to grow and get out of control.
Memory is consumed as scripts are loaded. Memory is consumed:

1. By the `server.lua_scripts` dictionary holding original text.
1. Internally by Lua to keep the compiled byte-code.

> So, If you have to use dynamic scripting, then just use plain EVAL, because it doesn't make sense to load them first.

Reviewer: just use the `EVAL` command

> So, If you have to use dynamic scripting, then just use plain EVAL, because it doesn't make sense to load them first.

Author: ```suggestion
So, If you have to use dynamic scripting, then just use the `EVAL` command, because it doesn't make sense to load them first.
```

#### Trade-offs

If you increase the `maxmemory-samples` configuration parameter, it expires the keys faster, but it costs more CPU cycles, which increases latency of commands. Secondly, increasing the expiry of keys helps but that requires significant changes to application logic.

### Use better serializer

Redis does not have any specific data type to store the serialized objects, they are stored as byte array in Redis. If you are using regular means of serializing our Java, Python, and PHP objects, they can be of larger size which impacts the memory consumption and latency.

#### Choosing a serializer

Instead of using the default serializer of your programming language (Java serialized objects, Python pickle, PHP serialize, and so on), switch to a better library like Protocol Buffers, MessagePack, and so on.

##### MessagePack

MessagePack is an efficient binary serialization format. It lets you exchange data among multiple languages like JSON. But it's faster and smaller. Small integers are encoded into a single byte, and typical short strings require only one extra byte in addition to the strings themselves.

> As stated by Salvatore Sanfilippo, creator of Redis, "Redis scripting has support for MessagePack because it is a fast and compact serialization format with a simple to implement specification. I liked it so much that I implemented a MessagePack C extension for Lua just to include it into Redis.""

Reviewer: Double "" in the end of the sentence.

> As stated by Salvatore Sanfilippo, creator of Redis, "Redis scripting has support for MessagePack because it is a fast and compact serialization format with a simple to implement specification. I liked it so much that I implemented a MessagePack C extension for Lua just to include it into Redis.""

Author: ```suggestion
As stated by Salvatore Sanfilippo, creator of Redis, "Redis scripting has support for MessagePack because it is a fast and compact serialization format with a simple to implement specification. I liked it so much that I implemented a MessagePack C extension for Lua just to include it into Redis."
```

##### Protocol buffers

> Protocol buffers, usually referred as Protobuf, is a protocol developed by Google to allow serialization and deserialization of structured data. Google developed it with the goal to provide a better way, compared to XML, to make systems communicate. So they focused on making it simpler, smaller, faster, and more maintainable then XML.

Reviewer: Add link - https://github.com/protocolbuffers/protobuf

> Protocol buffers, usually referred as Protobuf, is a protocol developed by Google to allow serialization and deserialization of structured data. Google developed it with the goal to provide a better way, compared to XML, to make systems communicate. So they focused on making it simpler, smaller, faster, and more maintainable then XML.

Author: ```suggestion
Protocol buffers, usually referred as [Protobuf](https://github.com/protocolbuffers/protobuf), is a protocol developed by Google to allow serialization and deserialization of structured data. Google developed it with the goal to provide a better way, compared to XML, to make systems communicate. So they focused on making it simpler, smaller, faster, and more maintainable then XML.
```

- Check if an observation already appeared in the stream.
- Find the fraction or the number of observations in the stream that are smaller or larger than a given value.

If you are ready to trade accuracy with speed and memory usage, consider using one of the following probabilistic data structures. You can use:

- [HyperLogLog](https://redis.io/docs/data-types/hyperloglogs/) to estimate the number of unique observations in a set.
- [Bloom filter or cuckoo filter](https://redis.io/docs/stack/bloom/) to check if an observation has already appeared in the stream. False positive matches are possible, but false negatives are not.
- [t-digest](https://redis.io/docs/stack/bloom/) to estimate the fraction or the number of observations in the stream that are smaller or larger than a given value.

Bloom filter and cuckoo filter require [RedisBloom](https://redis.com/modules/redis-bloom/).

#### Trade-offs

Following are the trade-offs of using HyperLogLog:

> 1. The results achieved from HyperLogLog are not 100% accurate; they have an approximate standard error of 0.81%.

Reviewer: remove 'approximate'
Consider simplifying:
The results achieved from HyperLogLog have a standard error of 0.81%.

> 1. The results achieved from HyperLogLog are not 100% accurate; they have an approximate standard error of 0.81%.

Author: ```suggestion
1. The results achieved from HyperLogLog have a standard error of 0.81%.
```

> 1. Hyperloglog only tells you the unique count. It cannot tell you the elements in the set.

Reviewer: can be used only for estimating the count of unique observations, and not for checking if a specific value was observed.

> 1. Hyperloglog only tells you the unique count. It cannot tell you the elements in the set.

Author: ```suggestion
1. Hyperloglogcan can be used only for estimating the count of unique observations, and not for checking if a specific value was observed.
```

For example, if you want to maintain how many unique IP addresses made an API call today, HyperLogLog tells you `46966 unique IPs for today`.

> But if your desired result is `Show me those 46966 IP Addresses`, it cannot show you. To achieve that, you need to maintain all IP addresses in a set.

Reviewer: Remove 'it cannot show you'.

But if you want to retrieve all the unique IP Addresses, you would need to maintain a set containing all these addresses.

> But if your desired result is `Show me those 46966 IP Addresses`, it cannot show you. To achieve that, you need to maintain all IP addresses in a set.

Author: ```suggestion
But if you want to retrieve all the unique IP Addresses, you would need to maintain a set containing all these addresses.
```

#### What does it mean to compress field names

Referring to the previous example of converting hashes to list, you had a hash with user details.

```bash
hmset user:123 id 123 firstname Bob lastname Lee
location CA twitter bob_lee
```

In this case, firstname, lastname, location, and twitter are field names that can been shortened to `fn`, `ln`, `loc`, and so on. By doing so, you can save some memory that is used by the field names.

### Compress values

Redis and clients are typically IO bound and the IO costs are typically at least 2 orders of magnitude in respect to the rest of the request/reply sequence. Redis by default does not compress any value that is stored in it. Therefore, you should compress your data before storing it in Redis. This helps to reduce the payload, which in return gives you higher throughput, lower latency, and higher cost savings.

#### How to compress strings

There are several compression algorithms to choose from, each with its own trade-offs.

1. [Snappy](https://en.wikipedia.org/wiki/Snappy_(compression)) aims for high speed and reasonable compression.
1. LZO compresses fast and decompresses faster.

> 1. Others, such as gzip, are more widely available.

Reviewer: Gzip

> The gzip compression uses more CPU resources than Snappy or LZO, but provides a higher compression ratio. gzip is often a good choice for cold data, which is accessed infrequently. Snappy or LZO are a better choice for hot data, which is accessed frequently.

Author: ```suggestion
The Gzip compression uses more CPU resources than Snappy or LZO, but provides a higher compression ratio. Gzip is often a good choice for cold data, which is accessed infrequently. Snappy or LZO are a better choice for hot data, which is accessed frequently.
```
